Allow using last (week|month|year) as date param

sgiehl opened the pull request:
### Description:

Using one those values actually won't directly pass them to strtotime, as it might have imho unexpected results. `last week` passed to `strtotime` for example always returns the monday of the previous week. Instead I guess it would be expected to result in the day 7 days ago. Similar applies to last month and year. 

For last month there are still some edge cases on the last day of a month when the previous month didn't have that much days.

E.g. `last month` on 31th Dezember would return 1st Dezember

fixes #13818

### Review

* [ ] Functional review done
* [x] Usability review done (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [x] Security review done [see checklist](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [x] Code review done
* [x] Tests were added if useful/possible
* [x] Reviewed for breaking changes
* [ ] Developer changelog updated if needed
* [ ] Documentation added if needed
* [ ] Existing documentation updated if needed

sgiehl committed:
Allow using last (week|month|year) as date param

sgiehl committed:
Adds some unit tests

diosmosis commented on tests/PHPUnit/Unit/DateTest.php:
```diff
@@ -72,6 +72,39 @@ public function testYesterday()
+    {
+        $date = Date::lastYear();
+        $this->assertEquals(strtotime(date("Y-m-d", strtotime('-1 year')) . " 00:00:00"), $date->getTimestamp());
+    }
```
I think you could test these more effectively by using `Date::$now`. Right now it looks like you're mostly duplicating the function logic in the tests?

sgiehl replied:
@diosmosis  I've improved the date tests (and also fixed `yesterdaySameTime`, that actually wasn't respecting the now timestamp for the returned time)

sgiehl committed:
improve tests

sgiehl committed:
Merge branch '4.x-dev' into dateformats

diosmosis commented:
@mattab / @tsteur would we need to update documentation anywhere for this?

diosmosis commented:
Should these be usable in the API? If so, would be good to have a system test or two for them I think.

diosmosis commented:
Tested locally, works well, tested comparison + export links. Left one other test related comment, otherwise lgtm

tsteur commented:
Yes be great to document here if it's supported in the API https://developer.matomo.org/api-reference/reporting-api

sgiehl committed:
Make it possible to use last (week|month|year) for api requests

sgiehl commented:
@diosmosis it actually wasn't possible to use those as api params. at least the results were incorrect.
I've added a new commit to make that possible. It also fixes the usage when period=range.
Also added some system tests.

diosmosis commented on tests/PHPUnit/System/SpecialDateParamsTest.php:
```diff
@@ -0,0 +1,78 @@
+                    'date'         => 'last year,last week',
+                    'periods'      => ['range'],
+                    'testSuffix'   => '_range',
+                ],
```
Just for completeness, maybe we could change a couple of these dates to `last-week/lastYear`?

sgiehl replied:
done

diosmosis reviewed (approved):
Left one more minor comment, otherwise LGTM

sgiehl committed:
improve test

sgiehl committed:
Merge branch '4.x-dev' into dateformats
